<img src="http://imgur.com/1ZcRyrc.png" style="float: left; margin: 20px; height: 55px"> 

## Lab: SQL Practice

_Author: Matt Brems_

### Required Preparation

In order to run SQL queries within Python/this Jupyter notebook, you will need to install the following:

```bash
conda install sqlalchemy
conda install psycopg2
```

If you get an error trying to install `psycopg2`, you can install a stand-alone package (sufficient for everything you'll need to complete this lab) by entering the following anywhere in the terminal: 

```bash 
pip install psycopg2-binary
```

(If `pip` doesn't work for you, try `pip3 install psycopg2-binary`.)

### Instructions

In this lab, you're going to have ten prompts. Each prompt will require you to do two things:
1. Write a SQL query.
2. Use the result of that query to answer the question.

The purpose of this lab is to get you to practice your SQL - an **important** skill in data science! While it is possible to get these answers using Pandas, it should be your goal to do everything only in SQL.

**Data**: The data used in this lab is the Iowa liquor database. It contains various tables of transactions. The schema for this database is here:

<img src="./images/schema.png" alt="schema" width="750"/>

In [1]:
# Import Pandas and Create_Engine

!pip install sqlalchemy
!pip install psycopg2

import pandas as pd
from sqlalchemy import create_engine

# Set up the engine to access the data.
engine = create_engine('postgresql://analytics_student:analyticsga@analyticsga-psql.generalassemb.ly:5432/iowa_liquor_sales_database')

### Prompt 1

How many rows are there in the `stores` table of the Iowa liquor database?

In [2]:
# Use a SQL query to find the answer to the above prompt.

sql = 'select * from stores'

df_stores = pd.read_sql_query(sql, engine)
df_stores

store                         name store_status  \
0      2106  Hillstreet News and Tobacco            A   
1      2112               Mike's Liquors            I   
2      2113               Jamboree Foods            A   
3      2119           Manly Liquor Store            I   
4      2130         Sycamore Convenience            A   
...     ...                          ...          ...   
1968   9917       Honey Creek Distillery            A   
1969   9918  Paradise Distilling Company            A   
1970   9919            Dehner Distillery            A   
1971   9920        Louisiana Spirits LLC            A   
1972   9921                  maDIKwe USA            A   

                                          store_address address_info  
0     2217 College\nCedar Falls, IA 506130000\n(42.5...         None  
1     407 Sharp St.\nGlenwood, IA 515340000\n(41.046...         None  
2     1119 Market St. Box 71\nGowrie, IA 505430000\n...         None  
3     133 East Main\nManly, IA 504560000\n(43.286863...         None  
4     617 Sycamore\nWaterloo, IA 507030000\n(42.4978...         None  
...                                                 ...          ...  
1968  29538, 210th St.\nKeosauqua, IA 52565\n(40.740...         None  
1969  245, Railroad Ave\nDubuque, IA 52001\n(42.4873...         None  
1970  7500, University Ave\nClive, IA\n(41.600352803...         None  
1971  20909, South I-10 Frontage Rd\nLacassine, IA 7...         None  
1972  3101, Ingersoll Ave\nDes Moines, IA 50312\n(41...         None  

[1973 rows x 5 columns]

**Prompt 1 Answer**: There are 1973 rows in stores table

### Prompt 2

If you sort the stores in alphabetical order by name, which store is fourth?

In [3]:
# Use a SQL query to find the answer to the above prompt.

sql = 'select * from stores \
       order by name asc \
       limit 4'

df_stores = pd.read_sql_query(sql, engine)
df_stores

store                                name store_status  \
0   4986  218 Fuel Express & Chubby's Liquor            A   
1   3047          3047 Cub Foods / Iowa City            I   
2   3060               3060 Cub Foods / Ames            I   
3   3061         3061 Cub Foods / Sioux City            I   

                                       store_address address_info  
0  68 Monroe St\nFloyd, IA 50435\n(43.12249790500...         None  
1  855 Highway 1 West\nIowa City, IA 522460000\n(...         None  
2  3121 Grand Ave\nAmes, IA 500100000\n(42.053458...         None  
3  1732 Hamilton Blvd\nSioux City, IA 511030000\n...         None

**Prompt 2 Answer**: When the stores are sorted in alphabetical order, '3061 Cub Foods / Sioux City' is fourth

### Prompt 3

Among those sales whose category name is `IMPORTED VODKA`, if you sort by vendor in alphabetical order, what is the bottle price and number of bottles bought in the first 3 transactions?
> You should give us three sets of numbers - one for each transaction.

In [4]:
# Use a SQL query to find the answer to the above prompt.

sql = 'select * from sales \
       limit 1'

pd.read_sql_query(sql, engine)

date convenience_store  store county_number county category  \
0 2014-11-13              None   2635            82  Scott  1032200   

           category_name vendor_no                            vendor   item  \
0  IMPORTED VODKA - MISC       370  Pernod Ricard USA/Austin Nichols  34036   

     description  pack  liter_size state_btl_cost btl_price  bottle_qty  total  
0  Absolut Pears    12         750         $11.49    $17.24           4  68.96

In [5]:
sql = "select * from sales \
       where category_name like 'IMPORTED VODKA'\
       order by vendor asc\
       limit 3"

df_sales = pd.read_sql_query(sql, engine)
df_sales

date convenience_store  store county_number county category  \
0 2014-07-01              None   4509            85  Story  1032080   
1 2014-06-04              None   4251            85  Story  1032080   
2 2014-06-04              None   4509            85  Story  1032080   

    category_name vendor_no                     vendor    item  \
0  IMPORTED VODKA        14  Adamba Imports Int'l Inc.  904783   
1  IMPORTED VODKA        10      A Hardy / U.S.A. Ltd.   36200   
2  IMPORTED VODKA        10      A Hardy / U.S.A. Ltd.   36200   

                   description  pack  liter_size state_btl_cost btl_price  \
0                 Vesica Vodka    12         750          $5.99     $8.99   
1  Michael Godard "Mr. Smooth"     6         750         $19.00    $28.50   
2  Michael Godard "Mr. Smooth"     6         750         $19.00    $28.50   

   bottle_qty   total  
0          12  107.88  
1           6  171.00  
2           6  171.00

The bottle price and bottle quantity bought in the first 3 transactions are ($8.99, 12) , ($28.50, 6) , ($28.50, 1) respectively.

**Prompt 3 Answer**: 


### Prompt 4

What is the total amount of liquor sold in February 2015?

In [6]:
# Use a SQL query to find the answer to the above prompt.

sql = "select sum(liter_size) from sales \
       where date between '2015-02-01' and '2015-02-28'"

pd.read_sql_query(sql, engine)

sum
0  145998203

**Prompt 4 Answer**:

### Prompt 5

Among all transactions where `IMPORTED VODKA` was sold, which vendor has the most transactions? Report the total number of transactions, the number of bottles sold, and the total amount of revenue from these transactions. Rename the columns to make sure there isn't any confusion as to what they mean!

In [7]:
# Use a SQL query to find the answer to the above prompt.

sql = "select vendor, count(*) as Number_of_transactions, \
       sum(bottle_qty) as Number_of_bottles, \
       sum(total) as Total_revenue from sales \
       where category_name like 'IMPORTED VODKA'\
       group by vendor \
       order by total_revenue desc"

pd.read_sql_query(sql, engine)

vendor  number_of_transactions  \
0       Pernod Ricard USA/Austin Nichols                   42338   
1                    Bacardi U.S.A. Inc.                   26441   
2                        Diageo Americas                   15725   
3        Constellation Wine Company Inc.                   12565   
4                        Jim Beam Brands                    5850   
5                            Stoli Group                    4443   
6                   Bacardi U.S.A., Inc.                    3344   
7                 Moet Hennessy USA Inc.                    2153   
8                   Imperial Brands Inc.                    2799   
9       Constellation Wine Company, Inc.                    1672   
10                               Proximo                    1628   
11                        Luxco-St Louis                    1128   
12                               MHW Ltd                    1159   
13                   Wilson Daniels Ltd.                     815   
14            Russian Standard Vodka USA                     415   
15                 Imperial Brands, Inc.                     426   
16                                 LUXCO                     140   
17               Moet Hennessy USA, Inc.                     256   
18                      Quadro Group LLC                     289   
19              Brown-Forman Corporation                     263   
20           WILLIAM GRANT AND SONS INC.                     232   
21                      Klin Spirits LLC                     218   
22  Shaw Ross International Importers LL                      10   
23      Western Spirits Beverage Co. LLC                      18   
24                 A Hardy / U.S.A. Ltd.                      33   
25           W. J. Deutsch And Sons Ltd.                      24   
26                     Jinro America Inc                      21   
27                                  None                      62   
28           Russian Standard Vodka, USA                      71   
29          WILLIAM GRANT AND SONS, INC.                      56   
30                     Wine Intelligence                       3   
31                             TY KU LLC                      17   
32            The Patron Spirits Company                       4   
33          Park Street Imports LLC Voli                      12   
34                      Sazerac Co. Inc.                      10   
35              Firestarter Spirits Inc.                      20   
36      Stanley Stawski Distributing Co.                       2   
37                 Ti Beverage Group Ltd                       2   
38        Charles Jacquin Et. Cie., Inc.                       2   
39                     Klin Spirits, LLC                       5   
40                     Phenix Brands LLC                       2   
41             Adamba Imports Int'l Inc.                       1   
42                A Hardy / U.S.A., Ltd.                       1   

    number_of_bottles  total_revenue  
0              403450     7880365.88  
1              236047     5771242.07  
2              152038     3346425.42  
3              116053     1895005.51  
4              104119     1807097.23  
5               35513      832071.19  
6               30239      765347.31  
7               10966      323126.54  
8               21033      265317.79  
9               12956      200283.64  
10               9431      178212.03  
11               8933      116914.19  
12               4493      109190.48  
13               2396       84246.72  
14               2421       40361.58  
15               3054       37950.20  
16               2374       32425.32  
17               1032       30783.53  
18               1054       25928.40  
19               1730       23337.88  
20               1119       20981.25  
21                784       15483.84  
22                516       13731.84  
23                484        9453.60  
24                320        9120.00  
25                768      

**Prompt 5 Answer**: 

### Prompt 6

The reason we sell things is to make money. In sales, the term **markup** means the amount of extra money charged for a product over the cost to make that product. (You can think of the markup as the profit for that product.)

In Iowa, the law states the minimum price at which liquor may be sold.

Calculate the markup (name this `Markup`) that shows the difference between the `btl_price` and the `state_btl_cost`. 

> If you haven't subtracted columns together before, [this StackOverflow question](https://stackoverflow.com/questions/7536996/multiplying-two-columns-in-sql-server/7537059) may be helpful.

In [8]:
# Use a SQL query to find the answer to the above prompt.

sql = 'select (btl_price - state_btl_cost) as Markup from sales'

pd.read_sql_query(sql, engine)

markup
0         $5.75
1         $5.75
2         $5.75
3         $8.58
4         $2.50
...         ...
3049908   $5.75
3049909  $10.00
3049910   $5.75
3049911   $5.75
3049912  $10.00

[3049913 rows x 1 columns]

### Prompt 7

Calculate the "Percentage Markup" by dividing markup by the state bottle cost. (Rather than the total profit per bottle, this will tell you how much stores are increasing their price over the state minimum.)

After calculating the "Percentage Markup," calculate the average percentage markup per **store**. Which store has the largest average markup, and what is that average markup?

In [9]:
# Use a SQL query to find the answer to the above prompt.

sql = 'select avg((btl_price - state_btl_cost) / state_btl_cost) as Average_Percentage_Markup, vendor from sales \
       group by vendor \
       order by Average_Percentage_Markup desc \
       limit 1'

pd.read_sql_query(sql, engine)

average_percentage_markup                    vendor
0                   0.539263  Ole Smoky Distillery LLC

**Prompt 7 Answer**: Ole Smoky Distillery LLC has the largest average markup at 53.9%

### Prompt 8

Building off of your last query, retrieve the five stores with the highest average markup percentage. Along with the store IDs and average markup percentage, please return the minimum and maximum markup percentage for those stores, as well as the names and addresses of those store.

What are the names of the five stores?

In [10]:
# Use a SQL query to find the answer to the above prompt.

sql = 'select avg((btl_price - state_btl_cost) / state_btl_cost) as Average_Percentage_Markup, \
       min((btl_price - state_btl_cost) / state_btl_cost) as Minimum_Percentage_Markup, \
       max((btl_price - state_btl_cost) / state_btl_cost) as Maximum_Percentage_Markup, vendor, vendor_no as store_IDs\
       from sales \
       group by vendor, vendor_no \
       order by Average_Percentage_Markup desc \
       limit 5'

pd.read_sql_query(sql, engine)

average_percentage_markup  minimum_percentage_markup  \
0                   0.539263                   0.500000   
1                   0.529461                   0.498670   
2                   0.525638                   0.500000   
3                   0.520482                   0.500000   
4                   0.516114                   0.499081   

   maximum_percentage_markup                       vendor store_ids  
0                   0.800720     Ole Smoky Distillery LLC       346  
1                   0.599800    Monsieur Henri Wines Ltd.       330  
2                   1.727520      Iowa Distilling Company       154  
3                   0.607066  Hood River Distillers, Inc.       971  
4                   0.587719               Gemini Spirits       330

**Prompt 8 Answer**: Ole Smoky Distillery LLC, 	Monsieur Henri Wines Ltd., Iowa Distilling Company, Hood River Distillers, Inc. and Gemini Spirits have the highest average markup percentage.

### Prompt 9

Which nine counties sell the highest number of bottles of liquor per capita (per person)? Calculate the average bottle size of the bottles of liquor sold by these counties rounded to the nearest first decimal place. Of the nine counties selling the highest number of bottles of liquor per capita, which county sells (on average) the largest bottle size?

**_BONUS FUN_**: This question can (also) be solved with a SQL subquery - that is, a query that queries a query! If you'd like to give this a try, you can read more about it [here](https://www.dofactory.com/sql/subquery). Otherwise, keep an eye out for the solution key!

In [11]:
# Use a SQL query to find the answer to the above prompt.

sql = 'select counties.county, counties.population, \
       round(sum(sales.bottle_qty) / counties.population, 1) as Number_bottles_of_liquor_per_capita \
       from counties \
       inner join sales on counties.county = sales.county \
       group by counties.county \
       order by Number_bottles_of_liquor_per_capita desc \
       limit 9'

pd.read_sql_query(sql, engine)

county  population  number_bottles_of_liquor_per_capita
0      Dickinson       16667                                 22.0
1           Polk      430640                                 15.0
2     Black Hawk      131090                                 15.0
3    Cerro Gordo       44151                                 14.0
4          Scott      165224                                 13.0
5        Johnson      130882                                 13.0
6           Linn      211226                                 12.0
7  Pottawattamie       93158                                 11.0
8        Kossuth       15543                                 11.0

**Prompt 9 Answer**: Dickinson on average sells the largest bottle size

### Prompt 10 (advanced but required!)

In the `sales` table, three columns are `btl_price`, `bottle_qty`, and `total`. The `total` for a transaction _should be_ the product of `btl_price` and `bottle_qty`. How many transactions have a value of `total` that is not equal to `btl_price` time `bottle_qty`?

> There will be a type error as well! We cannot compare type `money` to type `real`. We have to convert `total` to type `money` so that we can compare the two directly. [This link](https://docs.microsoft.com/en-us/sql/t-sql/functions/cast-and-convert-transact-sql?view=sql-server-ver15#syntax) may be helpful to you!

In [12]:
sql = 'select count(*) from sales \
       where cast(total as money) !=  (btl_price * bottle_qty)'

pd.read_sql_query(sql, engine)

count
0      0

In [13]:
sql = "select * from information_schema.columns \
       where table_name = 'sales'"

pd.read_sql_query(sql, engine)

table_catalog table_schema table_name        column_name  \
0   iowa_liquor_sales_database       public      sales               date   
1   iowa_liquor_sales_database       public      sales  convenience_store   
2   iowa_liquor_sales_database       public      sales              store   
3   iowa_liquor_sales_database       public      sales      county_number   
4   iowa_liquor_sales_database       public      sales             county   
5   iowa_liquor_sales_database       public      sales           category   
6   iowa_liquor_sales_database       public      sales      category_name   
7   iowa_liquor_sales_database       public      sales          vendor_no   
8   iowa_liquor_sales_database       public      sales             vendor   
9   iowa_liquor_sales_database       public      sales               item   
10  iowa_liquor_sales_database       public      sales        description   
11  iowa_liquor_sales_database       public      sales               pack   
12  iowa_liquor_sales_database       public      sales         liter_size   
13  iowa_liquor_sales_database       public      sales     state_btl_cost   
14  iowa_liquor_sales_database       public      sales          btl_price   
15  iowa_liquor_sales_database       public      sales         bottle_qty   
16  iowa_liquor_sales_database       public      sales              total   

    ordinal_position column_default is_nullable                    data_type  \
0                  1           None         YES  timestamp without time zone   
1                  2           None         YES                         text   
2                  3           None         YES                      integer   
3                  4           None         YES                         text   
4                  5           None         YES                         text   
5                  6           None         YES                         text   
6                  7           None         YES                         text   
7                  8           None         YES                         text   
8                  9           None         YES                         text   
9                 10           None         YES                      integer   
10                11           None         YES                         text   
11                12           None         YES                      integer   
12                13           None         YES                      integer   
13                14           None         YES                        money   
14                15           None         YES                        money   
15                16           None         YES                      integer   
16                17           None         YES                      numeric   

   character_maximum_length  character_octet_length  ...  is_identity  \
0                      None                     NaN  ...           NO   
1                      None            1.073742e+09  ...           NO   
2                      None                     NaN  ...           NO   
3                      None            1.073742e+09  ...           NO   
4                      None            1.073742e+09  ...           NO   
5                      None            1.073742e+09  ...           NO   
6                      None            1.073742e+09  ...           NO   
7                      None            1.073742e+09  ...           NO   
8                      None            1.073742e+09  ...           NO   
9                      None                     NaN  ...           NO   
10                     None            1.073742e+09  ...           NO   
11                     None                     NaN  ...           NO   
12                     None                     NaN  ...           NO   
13                     None                     NaN  ...           NO   
14                     None                     NaN  ...           NO   
15                   

**Prompt 10 Answer**: 0 Transactions have total value that is not equal to btl_price time bottle_qty